src/python/createrepo_c/__init__.py

```diff
@@ -382,6 +386,364 @@ def __init__(self, primary_path, filelists_path, other_path, newpkgcb=None, warn
+
+        packages = collections.OrderedDict()
+
+        xml_parse_primary(self._primary_xml_path, pkgcb=pkgcb, warningcb=warningcb, do_files=False)
+        if not only_primary:
+            xml_parse_filelists(self._filelists_xml_path, newpkgcb=newpkgcb, warningcb=warningcb)
+            xml_parse_other(self._other_xml_path, newpkgcb=newpkgcb, warningcb=warningcb)
+        return packages, warnings
+
+
+# both the path and the *XmlFile objects need to be tracked together because there's no way to get the path
+# back from the *XmlFile objects
+MetadataInfoHolder = collections.namedtuple("MetadataInfoHolder", ["path", "writer"])
+
+class RepositoryWriter:
+
+    _FINISHED_ERR_MSG = "Cannot perform action after the repository has already finished being written"
+
+    def __init__(self,
+                 destination,
+                 num_packages=None,
+                 unique_md_filenames=True,
+                 changelog_limit=10,
+                 compression=ZSTD_COMPRESSION,
+                 checksum_type=SHA256):
+
+        if changelog_limit:
+            assert isinstance(changelog_limit, int) and changelog_limit >= 0, "changelog_limit must be an integer >= 0"
+
+        self.repomd = Repomd()
+        self._destination_repo_path = Path(destination)
+
+        self._unique_md_filenames = unique_md_filenames
+        self._changelog_limit = changelog_limit
+        self._checksum_type = checksum_type
+
+        self._has_set_num_pkgs = False
+        self._finished = False
+
+        os.makedirs(self.path, exist_ok=True)
+        os.makedirs(self.repodata_dir, exist_ok=True)
+
+        def _compression_suffix(compressiontype):
+            suffix = compression_suffix(compressiontype)
+            return suffix if suffix else ""
+
+        self._compression = compression
+        self._compression_suffix = _compression_suffix(compression)
+
+        pri_xml_path = self.repodata_dir / ("primary.xml" + self._compression_suffix)
+        fil_xml_path = self.repodata_dir / ("filelists.xml" + self._compression_suffix)
+        oth_xml_path = self.repodata_dir / ("other.xml" + self._compression_suffix)
+
+        self.working_metadata_files = {
+            "primary": MetadataInfoHolder(
+                pri_xml_path, PrimaryXmlFile(str(pri_xml_path), compressiontype=compression)
+            ),
+            "filelists": MetadataInfoHolder(
+                fil_xml_path, FilelistsXmlFile(str(fil_xml_path), compressiontype=compression)
+            ),
+            "other": MetadataInfoHolder(
+                oth_xml_path, OtherXmlFile(str(oth_xml_path), compressiontype=compression)
+            ),
+        }
+        self.additional_metadata_files = {}
+
+        if num_packages is not None:
+            self.set_num_of_pkgs(num_packages)
+
+        self._updaterecords = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        if exc_type:
+            pass
```

Comment:
Also I think I probably should have made this raise any errors rather than swallowing them

Reply:
Actually maybe they get re-raised?

Reply:
I think they do, I tried this example:
```python
class Divide:
    def __init__(self, num1, num2):
        self.num1 = num1
        self.num2 = num2

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        pass

    def divide_by_zero(self):
        # can cause ZeroDivisionError exception
        print(self.num1 / self.num2)

# will raise a ZeroDivisionError
with Divide(3, 0) as r:
    r.divide_by_zero()
```
from https://www.geeksforgeeks.org/__exit__-in-python/ and the exception isn't swallowed.